addons/sales_team/models/crm_team_member.py

```diff
@@ -0,0 +1,188 @@
+            existing = self.env['crm.team.member'].search([('user_id', 'in', active.user_id.ids)])
+            user_mapping = dict.fromkeys(existing.user_id, self.env['crm.team'])
+            for membership in existing:
+                user_mapping[membership.user_id] |= membership.crm_team_id
+
+            for member in active:
+                teams = user_mapping.get(member.user_id, self.env['crm.team'])
+                remaining = teams - (member.crm_team_id | member._origin.crm_team_id)
+                if remaining:
+                    member.member_warning = _("Adding %(user_name)s in this team would remove him/her from its current teams %(team_names)s.",
+                                              user_name=member.user_id.name,
+                                              team_names=", ".join(remaining.mapped('name'))
+                                             )
+                else:
+                    member.member_warning = False
+
+
+    @api.model_create_multi
+    def create(self, values_list):
+        """ Specific behavior implemented on create
+
+          * mono membership mode: other user memberships are automatically
+            archived (a warning already told it in form view);
+          * creating a membership already existing as archived: do nothing as
+            people can manage them from specific menu "Members";
+        """
+        is_membership_multi = self.env['ir.config_parameter'].sudo().get_param('sales_team.membership_multi', False)
+        if is_membership_multi:
+            return super(CrmTeamMember, self).create(values_list)
+
+        self._synchronize_memberships(values_list)
+        return super(CrmTeamMember, self).create(values_list)
+
+    def write(self, values):
+        is_membership_multi = self.env['ir.config_parameter'].sudo().get_param('sales_team.membership_multi', False)
+        if not is_membership_multi and values.get('active'):
+            self._synchronize_memberships([
+                dict(user_id=membership.user_id.id, crm_team_id=membership.crm_team_id.id)
+                for membership in self
+            ])
+        return super(CrmTeamMember, self).write(values)
+
+    def _synchronize_memberships(self, user_team_ids):
+        """ Synchronize memberships: archive other memberships.
+
+        :param user_team_ids: list of pairs (user_id, crm_team_id)
+        """
+        existing = self.search([('user_id', 'in', [values['user_id'] for values in user_team_ids])])
+        user_memberships = dict.fromkeys(existing.user_id.ids, self.env['crm.team.member'])
+        for membership in existing:
+            user_memberships[membership.user_id.id] += membership
+
+        existing_to_archive = self.env['crm.team.member']
+        for values in user_team_ids:
+            to_archive = user_memberships.get(values['user_id'], self.env['crm.team.member']).filtered(
+                lambda m: m.active and m.crm_team_id.id != values['crm_team_id']
```

Comment:
Why not explicitly removing the unactive ones when searching then?

Reply:
Indeed. Previously this code did some things to unactive one (keeping only one, so unlinking duplicates archived members). Can be simplified now.